server/src/main/java/org/elasticsearch/action/admin/cluster/stats/CCSUsageTelemetry.java

```diff
@@ -123,12 +123,12 @@ public CCSUsageTelemetry() {
 
     public void updateUsage(CCSUsage ccsUsage) {
         assert ccsUsage.getRemotesCount() > 0 : "Expected at least one remote cluster in CCSUsage";
-        // TODO: fork this to a background thread? if yes, could just pass in the SearchResponse to parse it off the response thread
+        // TODO: fork this to a background thread?
         doUpdate(ccsUsage);
     }
 
-    // TODO: what is the best thread-safety model here? Start with locking model in order to get the functionality working.
-    private synchronized void doUpdate(CCSUsage ccsUsage) {
+    // This is not synchronized, instead we ensure that every metric in the class is thread-safe.
```

Comment:
I'm glad to see that org.HdrHistorgram has a ConcurrentHistogram that works nicely here. So the update (write) path looks thread safe.

However, the read/copy path needed for `getCCSTelemetrySnapshot` is not thread safe. The [javadoc](https://github.com/HdrHistogram/HdrHistogram/blob/master/src/main/java/org/HdrHistogram/ConcurrentHistogram.java) for ConcurrentHistogram says:

> ConcurrentHistogram provides no implicit synchronization that would prevent the contents of the histogram
> from changing during queries, iterations, copies, or addition operations on the histogram. Callers wishing to make
> potentially concurrent, multi-threaded updates that would safely work in the presence of queries, copies, or
> additions of histogram objects should either take care to externally synchronize and/or order their access,
> use the {@link SynchronizedHistogram} variant, or (recommended) use {@link Recorder} or
> {@link SingleWriterRecorder} which are intended for this purpose.

I'm concerned about their phrasing "safely work". What is the worst case scenario? Can it throw Exceptions? Will the data be potentially badly corrupted or significantly wrong? Or could it just be that some of the copies are slightly stale?

Slightly stale and misaligned histograms are probably OK for our use case where some fuzziness around latency precision is fine. But anything more than that would be problematic.

Is there a way we can either test or do some code auditing to see what the consequence is?

The safest option here would, IMO, be to use Java [ReadWriteLocks](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReadWriteLock.html), where we would actually use them "backwards". The updates would grab the read locks (which are quick) since the update paths are thread safe due to the data structures you've used. And the "copy" path (`getCCSTelemetrySnapshot`) would grab the write lock in order to get an exclusive lock that blocks all updates until it has finished copying.

I did some benchmarking of ReadWriteLocks at a previous company and obtaining the Read lock was very fast and did not impact throughput/latency for the use case I was testing.

Reply:
I did discuss r/w locks last Thursday and the team disposition was rather negative towards them, but that will certainly make it safer. 

I can try and make the test that would produce a read in parallel with a number of writes, to see if anything bad could happen there. 

It could also be an option to put r/w lock specifically on the Histogram (in LongMetric) instead of the whole thing. It'll have larger overhead but may reduce time-under-lock significantly.

Reply:
The idea of flipped r/w lock is interesting, but that'd mean we block parallel reads? Read performance may not be that important but if we lock the whole thing it may be a bit annoying? If we only lock an individual `LongMetric` it's probably not a big deal. But I'll think if it's possible to do better.

Reply:
Only the `getCCSTelemetrySnapshot` path would need to grab the write lock, which would block all telemetry updates from active, finishing cross-cluster searches, but I don't think there would be more than one thread calling  `getCCSTelemetrySnapshot` as in theory only Kibana calls the the cluster/stats endpoint with any regularity and only once per day per ES node.

Reply:
I took a look at ConcurrentHistogram, and looks like `resize()` operations have locks with other read operations, and the code here: 

https://github.com/HdrHistogram/HdrHistogram/blob/master/src/main/java/org/HdrHistogram/AbstractHistogram.java#L725-L771 makes special provisions to deal with ConcurrentHistogram using the locked methods. From that I make a conclusion that the worst case scenario when copying histogram while updating is that we lose some counts which are being counted while we're copying. Since it's OK with us, I _think_ it's ok to use it without locks.

I also made a test which runs a number of reads on the data constantly updated by a number of threads, and I do not see any anomalies. I verified that resizes did happen during the run. I can add this test to the suite if necessary?

Reply:
I also did some micro-benchmarking with adding a lock to `LongMetric` and on a single-threaded scenario it slows down `updateUsage` by about 25%. If I put the lock just on `doUpdate` the slowdown is about 8%.

Reply:
OK, thanks for the follow-up testing. I'll go ahead and approve as is, then.